### PR TITLE
Add a range argument to vec.extract_if

### PR DESCRIPTION
**Pull request by the8472**
tracking issue: #43244

This adds the range argument requested in https://github.com/rust-lang/rust/issues/43244#issuecomment-2486160659

### CODE REVIEW

File: library/alloc/src/vec/mod.rs [hunk 1]
```diff
@@ -3654,25 +3663,29 @@ impl<T, A: Allocator> Vec<T, A> {
...
     /// assert_eq!(evens, vec![2, 4, 6, 8, 14]);
     /// assert_eq!(odds, vec![1, 3, 5, 9, 11, 13, 15]);
     /// ```
+    ///
+    /// Using the range argument to only process a part of the vector:
+    ///
+    /// ```
+    /// #![feature(extract_if)]
+    /// let mut items = vec![0, 0, 0, 0, 0, 0, 0, 1, 2, 1, 2, 1, 2];
+    /// let ones = items.extract_if(7.., |x| *x == 1).collect::<Vec<_>>();
```
Comment (tgross35) [inline library/alloc/src/vec/mod.rs:3678]: Maybe change one of the first few `0`s to `1` to indicate they don't get extracted (out of range)?

Comment (the8472) [reply library/alloc/src/vec/mod.rs:3678]: Hrm, I wanted to show the "I have a prefix that I know doesn't contain those things, so I can skip them" use.

File: library/alloc/src/vec/extract_if.rs [hunk 2]
```diff
@@ -24,12 +24,13 @@ pub struct ExtractIf<
     T,
     F,
     #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global,
-> where
-    F: FnMut(&mut T) -> bool,
+>
 {
     pub(super) vec: &'a mut Vec<T, A>,
```
Comment (cuviper) [inline library/alloc/src/vec/extract_if.rs:29]: Do we still need all these fields to be `pub(super)` now that there's a local `fn new`?

File: library/alloc/src/vec/extract_if.rs [hunk 3]
```diff
@@ -82,14 +92,12 @@ where
...
     }
 }
 
 #[unstable(feature = "extract_if", reason = "recently added", issue = "43244")]
 impl<T, F, A: Allocator> Drop for ExtractIf<'_, T, F, A>
-where
-    F: FnMut(&mut T) -> bool,
 {
     fn drop(&mut self) {
         unsafe {
```
[the remark this answers is not in the document]
Comment (the8472) [reply library/alloc/src/vec/extract_if.rs:101]: removed the comment.

File: library/alloc/src/vec/extract_if.rs (hunk 2 above)
Comment (the8472) [reply library/alloc/src/vec/extract_if.rs]: removed the visibility.

File: library/alloc/tests/vec.rs [hunk 4]
```diff
@@ -1492,6 +1492,31 @@ fn extract_if_true() {
...
+    assert_eq!(count, 3);
+
+    let it = vec.extract_if(1..=3, |_| false);
+    assert_eq!(it.collect::<Vec<_>>(), vec![]);
+    assert_eq!(vec, vec![0, 4, 5, 6, 7, 8, 9, 10]);
+}
+
+#[test]
+#[should_panic]
+fn extraxt_if_out_of_bounds() {
```
Comment (cuviper) [inline library/alloc/tests/vec.rs:1515]:
```suggestion
fn extract_if_out_of_bounds() {
```

**Review by cuviper — APPROVED** [review]
LGTM.

The FCP finished -- do you want to go ahead and add the stabilization too?

**Comment by tgross35** [comment]
r? @cuviper (sorry, I didn't realize this was assigned to me)

> The FCP finished -- do you want to go ahead and add the stabilization too?

Since this is a pretty significant API and implementation change, maybe it would be good to hold off for a couple of weeks? In case anyone using this API unstably has feedback after the change (I doubt it, but doesn't hurt).

**Comment by the8472** [comment]
Changing and stabilizing in the same PR doesn't seem ideal to me either, especially when the change happened parallel to the FCP. I think it should bake a few days at least.

**Comment by cuviper** [comment]
I think you forgot the "extraxt" typo fix when you rebased.

**Comment by the8472** [comment]
Indeed, fixed.

@bors r=cuviper

File: library/alloc/src/vec/mod.rs [hunk 5]
```diff
@@ -3630,10 +3633,12 @@ impl<T, A: Allocator> Vec<T, A> {
     /// Using this method is equivalent to the following code:
     ///
     /// ```
+    /// # use std::cmp::min;
     /// # let some_predicate = |x: &mut i32| { *x == 2 || *x == 3 || *x == 6 };
     /// # let mut vec = vec![1, 2, 3, 4, 5, 6];
```
Comment (nk9) [inline library/alloc/src/vec/mod.rs:3638]: Apologies for the late comment, but I wanted to point out that this sample code doesn't behave the same way as the real `extract_if()`. Given the same predicate and range, if your vector were `[1, 2, 3, 3, 3, 3, 3, 3, 4, 5, 6]`, then all of the 3s would be removed. `i` is only incremented when an element is dropped, but `range.end` is unchanged, so the items shift down. I got very confused when reading the docs and trying to square this sample code with the explanation of how the function works.

Fortunately, the real `extract_if()` does not have this problem. Proposed change to the sample code:

```rust
    let some_predicate = |x: &mut i32| { *x == 2 || *x == 3 || *x == 5 };
    let mut vec = vec![0, 1, 2, 3, 4, 5, 6];
    let range = 1..5;
    let mut i = range.start;
    let mut end = range.end;
    while i < min(vec.len(), end) {
        if some_predicate(&mut vec[i]) {
            let _val = vec.remove(i);
            // your code here
            end -= 1;
        } else {
            i += 1;
        }
    }


    assert_eq!(vec, vec![0, 1, 4, 5, 6]);
```

**Review by nk9 — COMMENTED** [review]
I hope this is the right place to raise this question about the sample code in this PR.

File: library/alloc/src/vec/mod.rs (hunk 5 above)
Comment (the8472) [reply library/alloc/src/vec/mod.rs:3638]: If you want to propose a documentation improvement you can just open a PR.

**Pull request merged**